Lib/ast.py

```diff
@@ -63,7 +63,9 @@ def literal_eval(node_or_string):
     if isinstance(node_or_string, Expression):
         node_or_string = node_or_string.body
     def _raise_malformed_node(node):
-        raise ValueError(f'malformed node or string: {node!r}')
+        lno = getattr(node, 'lineno', None)
+        lno_str = f' on line {lno}' if lno else ''
+        raise ValueError(f'malformed node or string{lno_str}: {node!r}')
```

Comment:
Instead of creating temp variables and empty strings, can you try gradually constructing the message? Something like this
```py
message = "malformed node or string"
if lineno := getattr(node, 'lineno', None):
    message += f" on lineno {lineno}"
raise ValueError(message + f": {node!r}")
```